File: solutions/ruby/dnd-character/1/dnd_character.rb
# frozen_string_literal: true

# Character generator for the game Dungeons & Dragons
class DndCharacter
  BASE_HITPOINTS = 10

  attr_reader :strength, :dexterity, :constitution, :intelligence, :wisdom, :charisma, :hitpoints

  def self.modifier(constitution)
    ((constitution - 10) / 2)
  end

  def initialize
    @strength = ability
    @dexterity = ability
    @constitution = ability
    @intelligence = ability
    @wisdom = ability
    @charisma = ability
    @hitpoints = BASE_HITPOINTS + self.class.modifier(@constitution)
  end

  private

  def ability
    Random.rand(3..18)
  end
end

File: solutions/typescript/dnd-character/1/dnd-character.ts
const BASE_HITPOINTS = 10

export class DnDCharacter {
  readonly strength: number
  readonly dexterity: number
  readonly constitution: number
  readonly intelligence: number
  readonly wisdom: number
  readonly charisma: number
  readonly hitpoints: number

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore()
    this.dexterity = DnDCharacter.generateAbilityScore()
    this.constitution = DnDCharacter.generateAbilityScore()
    this.intelligence = DnDCharacter.generateAbilityScore()
    this.wisdom = DnDCharacter.generateAbilityScore()
    this.charisma = DnDCharacter.generateAbilityScore()
    this.hitpoints = BASE_HITPOINTS + DnDCharacter.getModifierFor(this.constitution)
  }

  public static generateAbilityScore(): number {
    const rolls = Array.from({ length: 4 }, () => 1 + Math.floor(Math.random() * 6))
    return rolls
      .sort((a, b) => a - b)
      .slice(1)
      .reduce((sum, val) => sum + val, 0)
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2)
  }
}
